File: Volunteer.java
import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.Set;

public class Volunteer extends User {
    private String educationLevel;
    private String major;
    private String contactInfo;
    private int totalRating;
    private int numOfRatings;
    private Set<Event> eventsJoined;
    private List<String> comments;

    // Constructor
    public Volunteer(String[] fullName, String educationLevel, String major, String contactInfo) {
        super(fullName);
        this.educationLevel = educationLevel;
        this.major = major;
        this.contactInfo = contactInfo;
        this.totalRating = 0;
        this.numOfRatings = 0;
        this.eventsJoined = new HashSet<>();
        this.comments = new ArrayList<>();
    }
    
    // Volunteers can sign up for events
    public void signUp(Event event) {
        if (!event.isFull()) {
            this.eventsJoined.add(event);
            event.addVolunteer(this);
        }
    }

    // Volunteers can cancel their event sign up
    public void cancelSignUp(Event event){
        event.removeVolunteer(this);
    }

    // Package-private visibility
    // Coordinators can only add ratings
    void addRating(int rating) {
        this.totalRating += rating;
        this.numOfRatings++;
    }

    // Coordinators can receive a volunteers avg rating
    double getAvgRating() {
        if (numOfRatings > 0) {
            return (double) totalRating / numOfRatings;
        }
        else {
            return 0;
        }
    }

    // Coordinators can only add comments
    void addComment(String comment) {
        if (comments.size() == 10) {
            comments.remove(0);
        }
        comments.add(comment);
    }

    // Coordinators can view comments
    List<String> getComments() {
        return new ArrayList<>(comments);
    }

    // Accessors
    public void setEducationalLvl(String educationLevel) {
        this.educationLevel = educationLevel;
    }

    public String getEducationalLvl() {
        return this.educationLevel;
    }

    public void setMajor(String major) {
        this.major = major;
    }

    public String getMajor() {
        return this.major;
    }

    public void setContactInfo(String contactInfo) {
        this.contactInfo = contactInfo;
    }

    public String getContactInfo() {
        return this.contactInfo;
    }

    public void setBio(String bio) {
        super.editBio(bio);
    }

    public String getBio() {
        return super.getBio();
    }
}

File: Volunteer.js
import User from './User';

class Volunteer extends User {
  
  constructor(fullName, educationLevel, major, contactInfo) {
    super(fullName);
    this.educationLevel = educationLevel;
    this.major = major;
    this.contactInfo = contactInfo;
    this.totalRating = 0;
    this.numOfRatings = 0;
    this.eventsJoined = new Set();
    this.comments = [];
  }

  // Volunteers can sign up for events
  signUp(event) {
    if (!event.isFull()) {
      this.eventsJoined.add(event);
      event.addVolunteer(this);
    }
  }

  // Volunteers can cancel their event sign up
  cancelSignUp(event) {
    event.removeVolunteer(this);
    this.eventsJoined.delete(event);
  }

  // Package-private visibility
  // Coordinators can only add ratings
  addRating(rating) {
    if (typeof rating === 'number' && rating > 0) {
      this.totalRating += rating;
      this.numOfRatings++;
    }
  }

  // Coordinators can receive a volunteers avg rating
  getAvgRating() {
    if (this.numOfRatings > 0) {
      return this.totalRating / this.numOfRatings;
    } else {
      return 0;
    }
  }

  // Coordinators can only add comments
  addComment(comment) {
    if (typeof comment === 'string' && comment.length <= 300) {
      if (this.comments.length >= 10) {
        this.comments.shift();
      }
      this.comments.push(comment);
    }
  }

  // Coordinators can view comments
  getComments() {
    return [...this.comments];
  }

  // Accessors
  setEducationalLvl(educationLevel) {
    this.educationLevel = educationLevel;
  }

  getEducationalLvl() {
    return this.educationLevel;
  }

  setMajor(major) {
    this.major = major;
  }

  getMajor() {
    return this.major;
  }

  setContactInfo(contactInfo) {
    this.contactInfo = contactInfo;
  }

  getContactInfo() {
    return this.contactInfo;
  }

  setBio(bio) {
    super.editBio(bio);
  }

  getBio() {
    return super.getBio();
  }
}
// module.exports = Volunteer;
export default Volunteer;
